### PR TITLE
table settings altering

### PR DESCRIPTION
**Pull request by pustovit**

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/SettingsAppliers.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 package io.crate.analyze;
 
 import com.google.common.base.Joiner;
+import com.sun.tools.classfile.Opcode;
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SettingsAppliers.java]: import not used

File: sql/src/main/java/io/crate/analyze/SettingsAppliers.java [hunk 2]
```diff
@@ -30,6 +31,7 @@
 import io.crate.types.DoubleType;
 import io.crate.types.FloatType;
 import io.crate.types.IntegerType;
+import javafx.scene.control.Tab;
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SettingsAppliers.java]: import not used

File: sql/src/main/java/io/crate/analyze/SettingsAppliers.java [hunk 3]
```diff
@@ -80,6 +82,25 @@ public IllegalArgumentException invalidException() {
         }
     }
 
+    public abstract static class BlockingSettingsApplier extends SettingsAppliers.AbstractSettingsApplier {
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SettingsAppliers.java]: ```...extends AbstractSettingsApplier``

File: sql/src/main/java/io/crate/analyze/SettingsAppliers.java [hunk 4]
```diff
@@ -80,6 +82,25 @@ public IllegalArgumentException invalidException() {
...
+
+        public static final Settings DEFAULT = ImmutableSettings.builder()
+                .put(TableParameterInfo.BLOCKS_READ, false)
+                .put(TableParameterInfo.BLOCKS_WRITE, false)
+                .put(TableParameterInfo.BLOCKS_METADATA, false)
+                .build();
+
+        public BlockingSettingsApplier(String name, Settings defaultSettings) {
+            super(name, defaultSettings);
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SettingsAppliers.java]: why not pass `DEFAULT` here?

File: sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java [hunk 5]
```diff
@@ -235,6 +253,116 @@ public void applyValue(ImmutableSettings.Builder settingsBuilder, Object value)
...
+        public WriteBlockSettingApplier() {
+            super(ES_TO_CRATE_SETTINGS_MAP.get(TableParameterInfo.BLOCKS_WRITE), DEFAULT);
+        }
+
+    }
+
+    private static class MetadataBlockSettingApplier extends SettingsAppliers.BlockingSettingsApplier {
+
+        public MetadataBlockSettingApplier() {
+            super(ES_TO_CRATE_SETTINGS_MAP.get(TableParameterInfo.BLOCKS_WRITE), DEFAULT);
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java]: 2nd argument not necessary

File: sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java [hunk 6]
```diff
@@ -235,6 +253,116 @@ public void applyValue(ImmutableSettings.Builder settingsBuilder, Object value)
...
+        public ReadBlockSettingApplier() {
+            super(ES_TO_CRATE_SETTINGS_MAP.get(TableParameterInfo.BLOCKS_READ), DEFAULT);
+        }
+
+    }
+
+    private static class WriteBlockSettingApplier extends SettingsAppliers.BlockingSettingsApplier {
+
+        public WriteBlockSettingApplier() {
+            super(ES_TO_CRATE_SETTINGS_MAP.get(TableParameterInfo.BLOCKS_WRITE), DEFAULT);
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java]: 2nd argument not necessary

File: sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java [hunk 7]
```diff
@@ -235,6 +253,116 @@ public void applyValue(ImmutableSettings.Builder settingsBuilder, Object value)
...
+
+            settingsBuilder.put(TableParameterInfo.FLUSH_DISABLE, ExpressionToObjectVisitor.convert(expression, parameters));
+
+        }
+    }
+
+    private static class ReadBlockSettingApplier extends SettingsAppliers.BlockingSettingsApplier {
+
+        public ReadBlockSettingApplier() {
+            super(ES_TO_CRATE_SETTINGS_MAP.get(TableParameterInfo.BLOCKS_READ), DEFAULT);
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java]: 2nd argument not necessary

File: sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java [hunk 8]
```diff
@@ -235,6 +253,116 @@ public void applyValue(ImmutableSettings.Builder settingsBuilder, Object value)
...
+            super(ES_TO_CRATE_SETTINGS_MAP.get(TableParameterInfo.FLUSH_DISABLE), DEFAULT);
+        }
+
+        public static final Settings DEFAULT = ImmutableSettings.builder()
+                .put(TableParameterInfo.FLUSH_DISABLE, false)
+                .build();
+
+        @Override
+        public void apply(ImmutableSettings.Builder settingsBuilder, Object[] parameters, Expression expression) {
+
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java]: remove empty line

File: app/src/main/dist/config/crate.yml [hunk 9]
```diff
@@ -694,7 +694,7 @@
 # insert it via SQL you might get inconsistent data when querying
 
 # enable/disable elasticsearch HTTP REST API
-# es.api.enabled: false
+ es.api.enabled: true
```
Comment (chaudum) [inline app/src/main/dist/config/crate.yml]: why? should be disabled by default.

Comment (pustovit) [inline app/src/main/dist/config/crate.yml]: UPS! of course it should, forgot to switch it off after testing

**Comment by mfelsche** [comment]
test create table as well

**Comment by mfelsche** [comment]
check correct validation for the new settings with the correct settingsapplier

**Comment by mfelsche** [comment]
test ALTER TABLE reset as well

File: sql/src/main/java/io/crate/metadata/settings/CrateTableSettings.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            return TableParameterInfo.TOTAL_SHARDS_PER_NODE;
+        }
+
+        @Override
+        public Integer defaultValue() {
+            return -1;
+        }
+    };
+
+    public static final StringSetting ROUTING_ALLOCATION_ENABLE = new StringSetting(Sets.newHashSet(
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/metadata/settings/CrateTableSettings.java]: use ImmutableSet.of(...)

File: sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java [hunk 11]
```diff
@@ -235,6 +263,95 @@ public void applyValue(ImmutableSettings.Builder settingsBuilder, Object value)
...
+        private RecoveryInitialShardsApplier() {
+            super(ES_TO_CRATE_SETTINGS_MAP.get(TableParameterInfo.RECOVERY_INITIAL_SHARDS), DEFAULT);
+        }
+
+        @Override
+        public void apply(ImmutableSettings.Builder settingsBuilder, Object[] parameters, Expression expression) {
+            Object shardsRecoverySettings;
+            try {
+                shardsRecoverySettings = ExpressionToObjectVisitor.convert(expression, parameters);
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java]: check for validation for int and string cases

File: sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java [hunk 12]
```diff
@@ -235,6 +263,95 @@ public void applyValue(ImmutableSettings.Builder settingsBuilder, Object value)
         }
     }
 
+    private static class ThresholdOpsSettingsApplier extends SettingsAppliers.AbstractSettingsApplier {
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/analyze/TablePropertiesAnalyzer.java]: use typed SettingsAppliers where appropriate

File: sql/src/main/java/io/crate/metadata/settings/CrateTableSettings.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        public String name() {
+            return TableParameterInfo.WARMER_ENABLED;
+        }
+
+        @Override
+        public Boolean defaultValue() {
+            return true;
+        }
+    };
+
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/metadata/settings/CrateTableSettings.java]: add the settings from TablePropertiesAnalyzer here as well

**Comment by mfelsche** [comment]
ack

**Pull request merged**